// ngx-scroll-event/dist/ngx-scroll-event/fesm2015/ngx-scroll-event.js
import { EventEmitter, Directive, Output, Input, HostListener, NgModule } from '@angular/core';

class NgxScrollEventDirective {
    constructor() {
        this.onscroll = new EventEmitter();
        this.bottomOffset = 100;
        this.topOffset = 100;
    }
    // handle host scroll
    scrolled($event) {
        this.elementScrollEvent($event);
    }
    // handle window scroll
    windowScrolled($event) {
        this.windowScrollEvent($event);
    }
    windowScrollEvent($event) {
        const target = $event.target;
        const scrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop || 0;
        const isReachingTop = scrollTop < this.topOffset;
        const isReachingBottom = (target.body.offsetHeight - (window.innerHeight + scrollTop)) < this.bottomOffset;
        const emitValue = { isReachingBottom, isReachingTop, originalEvent: $event, isWindowEvent: true };
        this.onscroll.emit(emitValue);
    }
    elementScrollEvent($event) {
        const target = $event.target;
        const scrollPosition = target.scrollHeight - target.scrollTop;
        const offsetHeight = target.offsetHeight;
        const isReachingTop = target.scrollTop < this.topOffset;
        const isReachingBottom = (scrollPosition - offsetHeight) < this.bottomOffset;
        const emitValue = { isReachingBottom, isReachingTop, originalEvent: $event, isWindowEvent: false };
        this.onscroll.emit(emitValue);
    }
}
NgxScrollEventDirective.decorators = [
    { type: Directive, args: [{
                selector: '[libScrollEvent]'
            },] }
];
NgxScrollEventDirective.ctorParameters = () => [];
NgxScrollEventDirective.propDecorators = {
    onscroll: [{ type: Output }],
    bottomOffset: [{ type: Input }],
    topOffset: [{ type: Input }],
    scrolled: [{ type: HostListener, args: ['scroll', ['$event'],] }],
    windowScrolled: [{ type: HostListener, args: ['window:scroll', ['$event'],] }]
};

class NgxScrollEventModule {
}
NgxScrollEventModule.decorators = [
    { type: NgModule, args: [{
                declarations: [NgxScrollEventDirective],
                imports: [],
                exports: [NgxScrollEventDirective]
            },] }
];

/*
 * Public API Surface of ngx-scroll-event
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgxScrollEventDirective, NgxScrollEventModule };
