// @ngx-lite/input-tag/esm2015/ngx-input-tag.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxInputTagComponent, formatter as tagFormatter, } from './ngx-input-tag.component';
import { NGX_INPUT_TAG_TAG_FORMATTER } from './ngx-input-tag.di-tokens';
import * as i0 from "@angular/core";
export class NgxInputTagModule {
    static forRoot(config = { tagFormatter }) {
        return {
            ngModule: NgxInputTagModule,
            providers: [
                {
                    provide: NGX_INPUT_TAG_TAG_FORMATTER,
                    useValue: config.tagFormatter,
                },
            ],
        };
    }
}
NgxInputTagModule.ɵmod = i0.ɵɵdefineNgModule({ type: NgxInputTagModule });
NgxInputTagModule.ɵinj = i0.ɵɵdefineInjector({ factory: function NgxInputTagModule_Factory(t) { return new (t || NgxInputTagModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(NgxInputTagModule, { declarations: [NgxInputTagComponent], imports: [CommonModule], exports: [NgxInputTagComponent] }); })();
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(NgxInputTagModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NgxInputTagComponent],
                exports: [NgxInputTagComponent],
            }]
    }], null, null); })();
